File: control-plane/api-gateway/src/routes/logs.rs
use axum::{
    extract::{Query, State},
    http::StatusCode,
    response::{IntoResponse, Json},
    routing::get,
    Router,
};
use entity::{entities::logs, Logs};
use sea_orm::{
    ColumnTrait, EntityTrait, Order, PaginatorTrait, QueryFilter, QueryOrder, QuerySelect,
};
use serde::{Deserialize, Serialize};
use serde_json::json;
use uuid::Uuid;

use crate::{auth::rbac::CanViewLogs, AppState};

const DEFAULT_LIMIT: u64 = 100;
const MAX_LIMIT: u64 = 500;

#[derive(Debug, Deserialize)]
pub struct LogsQuery {
    pub service: Option<String>,
    pub level: Option<String>,
    pub classification: Option<String>,
    pub agent_id: Option<Uuid>,
    pub workload_id: Option<Uuid>,
    pub organization_id: Option<Uuid>,
    pub from: Option<chrono::DateTime<chrono::Utc>>,
    pub to: Option<chrono::DateTime<chrono::Utc>>,
    pub q: Option<String>,
    pub limit: Option<u64>,
    pub offset: Option<u64>,
}

#[derive(Debug, Serialize)]
pub struct LogsResponse {
    pub entries: Vec<logs::Model>,
    pub total: u64,
    pub has_more: bool,
}

fn apply_filters(
    mut query: sea_orm::Select<Logs>,
    params: &LogsQuery,
) -> sea_orm::Select<Logs> {
    if let Some(service) = &params.service {
        query = query.filter(logs::Column::Service.eq(service));
    }
    if let Some(level) = &params.level {
        query = query.filter(logs::Column::Level.eq(level));
    }
    if let Some(classification) = &params.classification {
        query = query.filter(logs::Column::Classification.eq(classification));
    }
    if let Some(agent_id) = params.agent_id {
        query = query.filter(logs::Column::AgentId.eq(agent_id));
    }
    if let Some(workload_id) = params.workload_id {
        query = query.filter(logs::Column::WorkloadId.eq(workload_id));
    }
    if let Some(organization_id) = params.organization_id {
        query = query.filter(logs::Column::OrganizationId.eq(organization_id));
    }
    if let Some(from) = params.from {
        query = query.filter(logs::Column::CreatedAt.gte(from));
    }
    if let Some(to) = params.to {
        query = query.filter(logs::Column::CreatedAt.lte(to));
    }
    if let Some(q) = &params.q {
        query = query.filter(logs::Column::Message.contains(q));
    }
    query
}

pub async fn list_logs(
    CanViewLogs(_claims): CanViewLogs,
    State(state): State<AppState>,
    Query(params): Query<LogsQuery>,
) -> Result<impl IntoResponse, (StatusCode, Json<serde_json::Value>)> {
    let limit = params.limit.unwrap_or(DEFAULT_LIMIT).min(MAX_LIMIT);
    let offset = params.offset.unwrap_or(0);

    let base_query = apply_filters(Logs::find(), &params);

    let total = base_query
        .clone()
        .count(&state.db_conn)
        .await
        .map_err(|e| {
            tracing::error!(error = %e, "failed to count logs");
            (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(json!({ "error": "database error" })),
            )
        })?;

    let entries = base_query
        .order_by(logs::Column::CreatedAt, Order::Desc)
        .offset(offset)
        .limit(limit)
        .all(&state.db_conn)
        .await
        .map_err(|e| {
            tracing::error!(error = %e, "failed to list logs");
            (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(json!({ "error": "database error" })),
            )
        })?;

    let has_more = offset + (entries.len() as u64) < total;

    Ok((
        StatusCode::OK,
        Json(json!(LogsResponse {
            entries,
            total,
            has_more,
        })),
    ))
}

pub fn logs_routes() -> Router<AppState> {
    Router::new().route("/logs", get(list_logs))
}

File: app/src/lib/api/logs.ts
import { authedFetch } from './http';

const API_BASE = '/api';

export interface LogEntry {
    id: string;
    service: string;
    level: string;
    classification: string;
    message: string;
    agent_id: string | null;
    workload_id: string | null;
    organization_id: string | null;
    created_at: string;
}

export interface LogsResponse {
    entries: LogEntry[];
    total: number;
    has_more: boolean;
}

export interface LogsFilter {
    service?: string;
    level?: string;
    classification?: string;
    agent_id?: string;
    workload_id?: string;
    organization_id?: string;
    from?: string;
    to?: string;
    q?: string;
    limit?: number;
    offset?: number;
}

export interface LogsRetention {
    retention_days: number;
}

function buildQuery(filter: LogsFilter): string {
    const params = new URLSearchParams();
    for (const [key, value] of Object.entries(filter)) {
        if (value !== undefined && value !== null && value !== '') {
            params.set(key, String(value));
        }
    }
    const query = params.toString();
    return query ? `?${query}` : '';
}

export async function listLogs(token: string, filter: LogsFilter): Promise<LogsResponse> {
    const res = await authedFetch(`${API_BASE}/logs${buildQuery(filter)}`, {
        headers: { Authorization: `Bearer ${token}` },
    });
    if (!res.ok) throw new Error(`Failed to list logs: ${res.status}`);
    return res.json();
}

export async function getLogsRetention(token: string): Promise<LogsRetention> {
    const res = await authedFetch(`${API_BASE}/admin/settings/logs-retention`, {
        headers: { Authorization: `Bearer ${token}` },
    });
    if (!res.ok) throw new Error(`Failed to get logs retention: ${res.status}`);
    return res.json();
}

export async function updateLogsRetention(
    token: string,
    retentionDays: number,
): Promise<LogsRetention> {
    const res = await authedFetch(`${API_BASE}/admin/settings/logs-retention`, {
        method: 'PUT',
        headers: {
            Authorization: `Bearer ${token}`,
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({ retention_days: retentionDays }),
    });
    if (!res.ok) {
        const err = await res.json().catch(() => ({ error: res.status }));
        throw new Error(err.error ?? `Failed to update logs retention: ${res.status}`);
    }
    return res.json();
}
